internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go

```diff
@@ -0,0 +1,189 @@
+		{"Bad:install policy error", testInstallPolicyErr, nil},
+		{"Bad:create token error", nil, testCreateTokenErr},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// setup mock secretclient and expected return values for this test
+			secretClient := &mocks.SecretStoreClient{}
+
+			secretClient.On("InstallPolicy", testRootToken,
+				consulSecretsEngineOpsPolicyName, mock.Anything).
+				Return(test.installPolicyError).Once()
+
+			secretClient.On("CreateToken", testRootToken, mock.Anything).
+				Return(createTokenResult, test.createTokenError).Once()
+
+			token, _, err := NewWriter(lc, secretClient, flOpener).
+				CreateMgmtTokenForConsulSecretsEngine(testRootToken)
+
+			if test.installPolicyError != nil || test.createTokenError != nil {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				require.Equal(t, "createdTestToken", token["auth"].(map[string]interface{})["client_token"])
+				secretClient.AssertExpectations(t)
+			}
+		})
+	}
+}
+
+func TestCreateAndWriteForConsulSecretEngine(t *testing.T) {
+	lc, flOpener := setup()()
+
+	createTokenResult := getCreateTokenResultStub()
+
+	testInstallPolicyErr := errors.New("install policy error")
+	testCreateTokenErr := errors.New("create token error")
+	testTokenFileDir := "test"
+	testRootToken := "testRootToken"
+
+	tests := []struct {
+		name                 string
+		rootToken            string
+		installPolicyCallErr error
+		createTokenCallErr   error
+	}{
+		{"Ok:CreateAndWrite with client call ok", testRootToken, nil, nil},
+		{"Bad:CreateAndWrite with empty token", "", nil, nil},
+		{"Bad:CreateAndWrite with install policy call error", testRootToken, testInstallPolicyErr, nil},
+		{"Bad:CreateAndWrite with create token call error", testRootToken, nil, testCreateTokenErr},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// setup mock secretclient and expected return values for this test
+			secretClient := &mocks.SecretStoreClient{}
+
+			secretClient.On("InstallPolicy", testRootToken,
+				consulSecretsEngineOpsPolicyName, mock.Anything).
+				Return(test.installPolicyCallErr).Once()
+
+			secretClient.On("CreateToken", testRootToken, mock.Anything).
+				Return(createTokenResult, test.createTokenCallErr).Once()
+
+			testTokenFilePath := filepath.Join(testTokenFileDir,
+				"mgmt-token.json"+"_"+strconv.FormatInt(time.Now().UnixNano(), 10))
+			fileWriter := NewWriter(lc, secretClient, flOpener)
+			_, err := fileWriter.CreateAndWrite(test.rootToken, testTokenFilePath,
+				fileWriter.CreateMgmtTokenForConsulSecretsEngine)
+			defer func() {
+				_ = os.RemoveAll(filepath.Dir(testTokenFilePath))
+			}()
+
+			expectError := test.installPolicyCallErr != nil ||
+				test.createTokenCallErr != nil ||
+				len(test.rootToken) == 0
+
+			if expectError {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				require.FileExists(t, testTokenFilePath)
```

Comment:
I don't understand.  If we're using a mockable FileOpener, why is an actual file created?

Reply:
we are not using mockable fileopener here, concrete file creates